File: Data Structures & Algorithms/validate-parentheses/submission-1.py
class Solution:
    def isValid(self, s: str) -> bool:
        closed_bracket = {
            '}': '{',
            ']': '[',
            ')': '('
        }

        stack = []
        for b in s:
            if b in closed_bracket:
                if not stack or closed_bracket[b] != stack[-1]:
                    return False
                else:
                    stack.pop()
            else:
                stack.append(b)
        return len(stack) == 0

File: Data Structures & Algorithms/minimum-stack/submission-1.java
class MinStack {
    private Deque<Integer> stack;
    private Deque<Integer> minStack;

    public MinStack() {
        // use 2 deque / stack
        // 1 for maintaining the min, the other is a normal stack
        this.stack = new ArrayDeque<>();
        this.minStack = new ArrayDeque<>();
    }
    
    public void push(int val) {
        stack.push(val);

        if (minStack.size() != 0) {
            minStack.push(Math.min(minStack.peek(), val));
        } else {
            minStack.push(val);
        }
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}
